plugins/inputs/snmp_trap/snmp_trap.go

```diff
@@ -0,0 +1,147 @@
+package snmp_trap
+
+import (
+	"net"
+	"sync"
+	"time"
+
+	"github.com/influxdata/telegraf"
+	"github.com/influxdata/telegraf/plugins/inputs"
+	"github.com/influxdata/telegraf/plugins/inputs/snmp"
+
+	"github.com/soniah/gosnmp"
+)
+
+type SnmpTrap struct {
+	ServiceAddress string `toml:"service_address"`
+
+	acc      telegraf.Accumulator
+	listener *gosnmp.TrapListener
+	wg       sync.WaitGroup
+	timeFunc func() time.Time
+
+	makeHandlerWrapper func(func(packet *gosnmp.SnmpPacket, addr *net.UDPAddr)) func(packet *gosnmp.SnmpPacket, addr *net.UDPAddr)
+
+	Log telegraf.Logger
+}
+
+var sampleConfig = `
+  ## Local address and port to listen on.  Omit address to listen on
+  ## all interfaces.  Example "127.0.0.1:1234", default ":162"
+  #service_address = :162
+`
+
+func (s *SnmpTrap) SampleConfig() string {
+	return sampleConfig
+}
+
+func (s *SnmpTrap) Description() string {
+	return "Receive SNMP traps"
+}
+
+func (s *SnmpTrap) Gather(_ telegraf.Accumulator) error {
+	return nil
+}
+
+func init() {
+	inputs.Add("snmp_trap", func() telegraf.Input {
+		return &SnmpTrap{
+			timeFunc:       time.Now,
+			ServiceAddress: ":162",
+		}
+	})
+}
+
+func (s *SnmpTrap) Init() error {
+	return nil
+}
+
+func (s *SnmpTrap) Start(acc telegraf.Accumulator) error {
+	s.acc = acc
+	s.listener = gosnmp.NewTrapListener()
+	s.listener.OnNewTrap = makeTrapHandler(s)
+	s.listener.Params = gosnmp.Default
+
+	// wrap the handler, used in unit tests
+	if nil != s.makeHandlerWrapper {
+		s.listener.OnNewTrap = s.makeHandlerWrapper(s.listener.OnNewTrap)
+	}
+
+	s.wg.Add(1)
+	go func() {
+		defer s.wg.Done()
+
+		// no ip means listen on all interfaces, ipv4 and ipv6
+		err := s.listener.Listen(s.ServiceAddress)
+		if err != nil {
+			s.Log.Errorf("error in listen: %s", err)
+		}
+	}()
+
+	<-s.listener.Listening()
+
+	return nil
+}
+
+func (s *SnmpTrap) Stop() {
+	s.listener.Close()
+	s.wg.Wait()
+}
+
+func makeTrapHandler(s *SnmpTrap) func(packet *gosnmp.SnmpPacket, addr *net.UDPAddr) {
+	return func(packet *gosnmp.SnmpPacket, addr *net.UDPAddr) {
+		tm := s.timeFunc()
+		fields := map[string]interface{}{}
+		tags := map[string]string{}
+
+		tags["trap_version"] = packet.Version.String()
+		tags["source"] = addr.IP.String()
+
+		for _, v := range packet.Variables {
+			// build a name and value for each variable to use as tags
+			// and fields.  defaults are the uninterpreted values
+			name := v.Name
+			value := v.Value
+
+			// use system mibs to resolve the name if possible
+			_, _, oidText, _, err := snmp.SnmpTranslate(v.Name)
+			if nil == err {
```

Comment:
Wouldn't dropping it entirely make it harder to set up and debug?  How about always adding the oid in one tag and optionally adding other tags for mib/name if they're available?  Then even if someone doesn't have snmptranslate and net-snmp set up right, at least won't have missing data.

Reply:
It would be a bit harder to debug, you would need to go look at the logfile of the Telegraf instance.  We do have a internal metric for errors, so we could increment this (currently done by logging an error or using acc.AddError).

I still think we should skip the metrics in spite of this, because there isn't any way to clean up bad data in the database.  Once it has been added, it is essentially permanent until and unless the retention policy removes it.

Reply:
> Once it has been added, it is essentially permanent until and unless the retention policy removes it.

I guess this isn't completely true, you could always use:
```
drop series from snmp_tap where 'oid' == '0.1.2.3'
```

Reply:
I'll change it to drop the trap and log a warning.  I agree putting data into the database with numeric